### DISCUSSION TITLE
nvim functions inside keymaps

### DISCUSSION THREAD

**Discussion by antinomie8** [neovim/neovim#30573] [Q&A] [answered]
I have the following keymap in my config files : 
`vim.keymap.set("n", "<leader>dbg", ":!clang++ --debug " .. vim.fn.fnamemodify(vim.fn.expand("%"), ":t") .. " -o " .. vim.fn.fnamemodify(vim.fn.expand("%"), ":t:r") .. ".exe<CR>")
`
When I trie to use it, only `:clang++ --debug  -o  .exe` is printed to the command line and obviously clang throws an error.
What I would like this keymap to do is use the result of vim.fn.fnamemodify as a string and concatenate it with the others to form a working keymap. 

I have tried wrapping it into a function, using `{expr=true}` and `{noremap=true}` but none of them works.

Does anyone know how to do that ?

**Comment by zeertzjq** [comment, marked as answer]
```lua
vim.keymap.set("n", "<leader>dbg", function()
  return ":!clang++ --debug " .. vim.fn.fnamemodify(vim.fn.expand("%"), ":t") .. " -o " .. vim.fn.fnamemodify(vim.fn.expand("%"), ":t:r") .. ".exe<CR>"
end, { expr = true })
```
